components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTSignatureValidationUtils.java

```diff
@@ -0,0 +1,276 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2.util;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.crypto.RSASSAVerifier;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.IdentityProviderProperty;
+import org.wso2.carbon.identity.application.common.util.IdentityApplicationManagementUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2ClientException;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2ServerException;
+import org.wso2.carbon.identity.oauth2.validators.jwt.JWKSBasedJWTValidator;
+
+import java.security.PublicKey;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+
+/**
+ * Utility class for JWT Signature Validation related functionalities.
+ */
+public class JWTSignatureValidationUtils {
+
+    private static final Log log = LogFactory.getLog(JWTSignatureValidationUtils.class);
+
+    private static final String JWKS_URI = "jwksUri";
+    private static final String JWKS_VALIDATION_ENABLE_CONFIG = "JWTValidatorConfigs.Enable";
+    private static final String ENFORCE_CERTIFICATE_VALIDITY
+            = "JWTValidatorConfigs.EnforceCertificateExpiryTimeValidity";
+
+    /**
+     * Method to validate the signature of the JWT.
+     *
+     * @param signedJWT signed JWT whose signature is to be verified.
+     * @param idp       Identity provider who issued the signed JWT.
+     * @return whether signature is valid, true if valid else false.
+     * @throws JOSEException
+     * @throws IdentityOAuth2Exception
+     */
+    public static boolean validateSignature(SignedJWT signedJWT, IdentityProvider idp)
+            throws JOSEException, IdentityOAuth2Exception {
+
+        boolean isJWKSEnabled = false;
+        String jwksUri = null;
+
+        isJWKSEnabled = isJWKSEnabled();
+
+        jwksUri = getJWKSUri(idp);
+
+        if (isJWKSEnabled && (jwksUri != null)) {
+            return validateUsingJWKSUri(signedJWT, jwksUri);
+        } else {
+            return validateUsingCertificate(signedJWT, idp);
+        }
+    }
+
+    /**
+     * Method to check whether the JWKS is enabled.
+     *
+     * @return boolean value depending on whether the JWKS is enabled.
+     */
+    private static boolean isJWKSEnabled() {
```

Comment:
isJWKSSignatureValidationEnabled() or something that indicates whether the JWKS based signature validation is enabled.

Please check whether we need to do this check at all.

Reply:
Is this check added to have some back-ward compatibility in other flows?

Reply:
It is checking whether the JWKS based signature validation is enabled. We can remove it, as long as there won't be a JWKS uri set while JWKS based signature validation is disabled.